testsuite/MDAnalysisTests/lib/test_cutil.py

```diff
@@ -45,6 +45,47 @@ def test_unique_int_1d(values):
     assert type(res) == type(ref)
     assert res.dtype == ref.dtype
 
+@pytest.mark.parametrize('arrays', (
+    [],  # empty array
+    [1, 1, 1, 1, ],  # all identical
+    [2, 3, 5, 7, ],  # unique, sorted
+    [5, 2, 7, 3, ],  # unique, unsorted
+    [1, 2, 2, 4, 4, 6, ],  # duplicates, sorted
+    [1, 2, 2, 6, 4, 4, ],  # duplicates, unsorted
+))
+def test_inverse_unique_contiguous_1d_array(arrays):
+    # Generate unique, unsorted arrays.
+    array = np.array(arrays, dtype=np.intp)
+    values, indices = np.unique(array, return_index=True)
+    unique_array = array[np.sort(indices)]
```

Comment:
I'm having a hard time understanding why you are doing a sort operation here. `unique_array`, as defined for the Cython version of `inverse_unique_contiguous_1d_array` is (emphasis mine):

> the unique, **unsorted** values of full_arr.

So, shouldn't it work whether the values are sorted or not?

For example, this diff still passes the tests and is a bit easier for me to parse:

```diff
--- a/testsuite/MDAnalysisTests/lib/test_cutil.py
+++ b/testsuite/MDAnalysisTests/lib/test_cutil.py
@@ -57,9 +57,11 @@ def test_inverse_unique_contiguous_1d_array(arrays):
     # Generate unique, unsorted arrays.
     array = np.array(arrays, dtype=np.intp)
     values, indices = np.unique(array, return_index=True)
-    unique_array = array[np.sort(indices)]
+    unique_array = array[indices]
     
     inverse = inverse_unique_contiguous_1d_array(array, unique_array)
+    expected_inverse = np.unique(array, return_inverse=True)[1]
+    assert_equal(inverse, expected_inverse)
     ref = array
     res = unique_array[inverse]
     assert_equal(res, ref)
@@ -71,7 +73,7 @@ def test_inverse_unique_contiguous_1d_array(arrays):
         # change array shape
         array_2d = array.reshape((2,-1))
         values, indices = np.unique(array_2d, return_index=True)
-        unique_array = array[np.sort(indices)]
+        unique_array = array[indices]
         inverse_unique_contiguous_1d_array(array_2d, unique_array)
     
     # raise ValueError on non-contiguous data
@@ -79,7 +81,7 @@ def test_inverse_unique_contiguous_1d_array(arrays):
     if not array.data.contiguous: # empty array ([]) is still contiguous
       with pytest.raises(ValueError) as e_info:
           values, indices = np.unique(array, return_index=True)
-          unique_array = array[np.sort(indices)]
+          unique_array = array[indices]
           inverse = inverse_unique_contiguous_1d_array(array, unique_array)
```

I'm also not really sure if we need the block with:

```python
     ref = array
     res = unique_array[inverse]
     assert_equal(res, ref)
```

It might be sufficient to confirm that the MDAnalysis version of `inverse` matches the same functionality in NumPy (though this is more minor point I suppose).

Reply:
This hasn't been changed by me and was taken directly from the function which generates the unique indicies just prior to the change:
https://github.com/MDAnalysis/mdanalysis/blob/800298ad271a00762ba3f51629c1ce41b7bf4172/package/MDAnalysis/lib/util.py#L233-L235

As far as I understand it, np.unique returns the array **sorted** and the return_index returns the sorted array inverse, by sorting the sorted unique array by the inverse you then return the **unsorted**, unique array.
Now that I type that out I suppose the proper thing to do is to import that function to use, then any changes to it will also remain in the tests.

This new cython code then generates the inverse of the new unsorted array.

Now that I type that out I suppose the proper thing to do is to import that function to use, then any changes to that function will also be expressed in the tests.

Perhaps this might also be a target for future improvement to remove the redundent sorts (and this code) by implimenting a unique function that returns the unique array and its inverse without the two additional sorts (sorting in np.unique, sorting to unsorted the results).

There already exists a cython 1d unique function, but that relies on sorting first too:
https://github.com/MDAnalysis/mdanalysis/blob/800298ad271a00762ba3f51629c1ce41b7bf4172/package/MDAnalysis/lib/_cutil.pyx#L54-L57